android/expoview/src/main/java/host/exp/exponent/fcm/ExpoFcmMessagingService.java

```diff
@@ -40,44 +39,36 @@ public void onMessageReceived(RemoteMessage remoteMessage) {
-
-          // If an experience is on SDK newer than 39, that is SDK40 and beyond up till UNVERSIONED
-          // we only use the new notifications API as it is going to be removed from SDK40.
-          if (sdkVersion >= 40) {
-            dispatchToNextNotificationModule(remoteMessage);
-            return;
-          } else if (sdkVersion == 38 || sdkVersion == 39) {
-            // In SDK38 and 39 we want to let people decide which notifications API to use,
-            // the next or the legacy one.
-            JSONObject androidSection = manifest.optJSONObject("android");
-            if (androidSection != null) {
-              boolean useNextNotificationsApi = androidSection.optBoolean("useNextNotificationsApi", false);
-              if (useNextNotificationsApi) {
-                dispatchToNextNotificationModule(remoteMessage);
-                return;
-              }
+    ExperienceDBObject experienceDBObject = ExponentDB.experienceIdToExperienceSync(remoteMessage.getData().get("experienceId"));
+    if (experienceDBObject != null) {
+      try {
+        JSONObject manifest = new JSONObject(experienceDBObject.manifest);
+        int sdkVersion = ABIVersion.toNumber(manifest.getString("sdkVersion")) / 10000;
+
+        // If an experience is on SDK newer than 39, that is SDK40 and beyond up till UNVERSIONED
+        // we only use the new notifications API as it is going to be removed from SDK40.
+        if (sdkVersion >= 40) {
+          dispatchToNextNotificationModule(remoteMessage);
+          return;
+        } else if (sdkVersion == 38 || sdkVersion == 39) {
+          // In SDK38 and 39 we want to let people decide which notifications API to use,
+          // the next or the legacy one.
+          JSONObject androidSection = manifest.optJSONObject("android");
+          if (androidSection != null) {
+            boolean useNextNotificationsApi = androidSection.optBoolean("useNextNotificationsApi", false);
+            if (useNextNotificationsApi) {
+              dispatchToNextNotificationModule(remoteMessage);
+              return;
             }
           }
-          // If it's an older experience or useNextNotificationsApi is set to false, let's use the legacy notifications API
-          dispatchToLegacyNotificationModule(remoteMessage);
-        } catch (JSONException e) {
-          e.printStackTrace();
-          onFailure();
         }
-      }
-
-      @Override
-      public void onFailure() {
-        Log.w("expo-notifications", "Couldn't get experience from remote message.", null);
+        // If it's an older experience or useNextNotificationsApi is set to false, let's use the legacy notifications API
         dispatchToLegacyNotificationModule(remoteMessage);
+      } catch (JSONException e) {
+        e.printStackTrace();
+        EXL.e("expo-notifications", "No experience found for id " + remoteMessage.getData().get("experienceId"));
```

Comment:
Actually, don't we guard against `No experience found` situation with `if (experienceDBObject != null)`? I mean… if we are inside the `if` doesn't it mean an experience has been found?